File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int func(vector<int>& bloomDay,int k,int i){
        int res=0;
        int count=0;
        for(int j=0;j<bloomDay.size();j++){
            if(bloomDay[j]>i) {
            res+=count/k;
            count=0;}
            if(bloomDay[j]<=i){
                count++;
            }
        }
        res+=count/k;
        return res;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
    if(((long)m*k)>bloomDay.size())return -1;
    int low=*min_element(bloomDay.begin(),bloomDay.end());
    int high=*max_element(bloomDay.begin(),bloomDay.end());
    int ans=high;
    while(low<=high){
        int mid=(low+high)/2;
        int res=func(bloomDay,k,mid);
        if(res>=m){
            if(mid<ans)ans=mid;
            high=mid-1;
        }
        else{
            low=mid+1;
        }
    }
    return ans;
    }
};

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def check_bouquet(self,min_days,bloomDay,m,k):
        count=0
        for i in range(0,len(bloomDay)):
            if(bloomDay[i]<=min_days):
                count+=1
                if(count==k):
                    count=0
                    m-=1
                    if(m==0):
                        return True
            else:
                count=0
        return False
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        if(m*k>len(bloomDay)):
            return -1
        low=min(bloomDay)
        high=max(bloomDay)
        res=-1
        while(low<=high):
            mid=(low+high)//2
            if(self.check_bouquet(mid,bloomDay,m,k)):
                res=mid
                high=mid-1
            else:
                low=mid+1
        return res